### PR TITLE
[MI-2341][webapp+server] Added filters for "Build" and "Release" events

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
1. Added "Release Pipeline", "Approval Status", "Approval Type", "Stage Name" and "Status" for release events. 
3. Added "Build Pipeline" and "Build Status" for build events. 
4. 3. Updated the API to get filter values for above events.

### CODE REVIEW

File: server/constants/constants.go [hunk 1]
```diff
@@ -92,12 +92,15 @@ const (
...
-	FilterBoards          = "boards"
-	FilterRepos           = "repos"
+	FilterCreatedByMe            = "me"
+	FilterCreatedByAnyone        = "anyone"
+	FilterAllChannels            = "all_channels"
+	FilterAll                    = "all"
+	FilterBoards                 = "boards"
+	FilterRepos                  = "repos"
+	EventTypeRelease             = "release"
+	FilterIDReleaseDefinitionID  = "releaseDefinitionId"
```
Comment (Kshitij-Katiyar) [inline server/constants/constants.go:102]: Is 2 times ID is needed in these variables ?

File: webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    handleSelectApprovalStatus,
+    selectedApprovalStatus,
+    selectedReleaseStatus,
+    handleSelectReleaseStatus,
+}: PipelinesFilterProps) => {
+    const {buildStatusOptions, releaseApprovalTypeOptions, releaseApprovalStatusOptions, releaseStatusOptions, subscriptionFiltersForPipelines, subscriptionFiltersNameForPipelines} = pluginConstants.form;
+
+    const {
+        getApiState,
+        makeApiRequestWithCompletionStatus,
```
Comment (Kshitij-Katiyar) [inline webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:57]: can this be in 1 line

File: webapp/src/components/card/subscription/index.tsx [hunk 3]
```diff
@@ -27,7 +27,7 @@ const SubscriptionCard = ({handleDeleteSubscrption, subscriptionDetails: {channe
 
     return (
         <BaseCard>
-            <>
+            <div>
```
Comment (Kshitij-Katiyar) [inline webapp/src/components/card/subscription/index.tsx:30]: why this change ?

File: webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+                                loadingOptions={isLoading}
+                                disabled={isLoading}
+                            />
+                        </div>
+                    </>
+                )
+            }
+            {
+                eventType === pluginConstants.common.eventTypePipelineKeys.releaseDeploymentApprovalComplete && (
+                    <>
```
Comment (Kshitij-Katiyar) [inline webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:179]: i think there is no need for these just the below div is enough

File: webapp/src/pluginConstants/form.ts [hunk 5]
```diff
@@ -105,6 +105,78 @@ const serviceTypeOptions: LabelValuePair[] = [
...
+    {
+        value: 'Failed',
+        label: 'Failed',
+    },
+    {
+        value: 'Stopped',
+        label: 'Stopped',
+    },
+    {
+        ...filterLabelValuePairAll,
```
Comment (Kshitij-Katiyar) [inline webapp/src/pluginConstants/form.ts:126]: can we change this filterLabelValuePairAll variable name ?

File: server/constants/constants.go (hunk 1 above)
Comment (raghavaggarwal2308) [reply server/constants/constants.go:102]: Yes, the first part is prefix and the second one is the name

File: webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    organization: string
+    projectId: string
+    eventType: string
+    selectedBuildPipeline: string
+    handleSelectBuildPipeline: (value: string, name?: string) => void
+    setIsFiltersError: (value: boolean) => void
+    selectedBuildStatus: string
+    handleSelectBuildStatus: (value: string, name?: string) => void
+    handleSelectReleasePipeline: (value: string, name?: string) => void
+    selectedReleasePipeline: string
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:22]:
```suggestion
    handleSelectReleasePipeline: (value: string, name?: string) => void
    selectReleasePipeline: string
```

Comment (ayusht2810) [reply webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:22]: shouldn't this be like this or vice-versa? if yes, check everywhere

File: webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        makeApiRequestWithCompletionStatus,
+    } = usePluginApi();
+
+    const getSubscriptionFiltersRequest = useMemo<GetSubscriptionFiltersRequest>(() => ({
+        organization,
+        projectId,
+        filters: subscriptionFiltersForPipelines,
+        eventType,
+        releasePipelineId: selectedReleasePipeline,
+    }), [organization, projectId, eventType, subscriptionFiltersForPipelines, selectedBuildPipeline, selectedReleasePipeline]);
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:66]: is this `selectedBuildPipeline` required?

File: webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+                getSubscriptionFiltersRequest,
+            );
+        }
+    }, [getSubscriptionFiltersRequest]);
+
+    const {data, isLoading, isError, isSuccess} = getApiState(
+        pluginConstants.pluginApiServiceConfigs.getSubscriptionFilters.apiServiceName,
+        getSubscriptionFiltersRequest as APIRequestPayload,
+    );
+    const filtersData = data as GetSubscriptionFiltersResponse || [];
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:81]:
```suggestion
    );
    
    const filtersData = data as GetSubscriptionFiltersResponse || [];
```

File: webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+
+    useEffect(() => {
+        if (isError && !isSuccess) {
+            setIsFiltersError(true);
+        } else {
+            setIsFiltersError(false);
+        }
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:88]:
```suggestion
            setIsFiltersError(true);
            return;
        } 
        
        setIsFiltersError(false);
        }
```

File: webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    }, [isLoading, isError, isSuccess]);
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:89]: is `isLoading` required here?

File: webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+                                loadingOptions={isLoading}
+                                disabled={isLoading}
+                            />
+                        </div>
+                    </>
+                )
+            }
+            {
+                eventType.includes('release') && (
+                    <>
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:127]:
```suggestion
```

Comment (ayusht2810) [reply webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:127]: check everywhere

File: webapp/src/containers/modals/SubscribeModal/index.tsx [hunk 12]
```diff
@@ -282,6 +283,48 @@ const SubscribeModal = () => {
             areaPath: newValue,
         });
 
+    const handleSetBuildPipelineFilter = (newValue: string, pipelineName?: string) =>
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/index.tsx:286]: `pipelineName` this is not being used anywhere, is this correct?
same for below as well

File: webapp/src/pluginConstants/form.ts (hunk 5 above)
Comment (raghavaggarwal2308) [reply webapp/src/pluginConstants/form.ts:126]: I think what we have is fine. Do you have any suggestions?

File: webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx (hunk 6 above)
Comment (raghavaggarwal2308) [reply webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:22]: No, we are storing the selected value in this variable and `handleSelectReleasePipeline` is used when we select a value from the dropdowns

File: webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx (hunk 10 above)
Comment (raghavaggarwal2308) [reply webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:89]: I think it is used to check for error when loading is completed. Right @avas27JTG ?

File: server/constants/constants.go (hunk 1 above)
Comment (Kshitij-Katiyar) [reply server/constants/constants.go:102]: ok then i think the last ID is not needed like for the filter for boards we have filterBoards not filterBoardsFilter

File: webapp/src/pluginConstants/form.ts (hunk 5 above)
Comment (Kshitij-Katiyar) [reply webapp/src/pluginConstants/form.ts:126]: can you put all in between the words like filterLabelPairAllValue or filterLabelPairValues. if that doesn't change the meaning

**Review by Kshitij-Katiyar — APPROVED** [review]

File: webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx (hunk 10 above)
Comment (ayusht2810) [reply webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:88]: on completion of loading, either the `isError` or `isSuccess` will change. So, I don't think we require that

**Review by ayusht2810 — APPROVED** [review]

File: webapp/src/pluginConstants/form.ts (hunk 5 above)
Comment (raghavaggarwal2308) [reply webapp/src/pluginConstants/form.ts:126]: Yes, it changes the meaning. The meaning of the name is that it's a label-value pair that contains the value "all"

File: server/constants/constants.go (hunk 1 above)
Comment (raghavaggarwal2308) [reply server/constants/constants.go]: Done

File: webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx (hunk 10 above)
Comment (avas27JTG) [reply webapp/src/containers/modals/SubscribeModal/filters/pipelines.tsx:88]: yes we can remove `isLoading`

**Review by avas27JTG — APPROVED** [review]

**Review by ayusht2810 — APPROVED** [review]
